flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayElementArgumentTypeStrategy.java

```diff
@@ -31,24 +31,28 @@
 
 import java.util.Optional;
 
-import static org.apache.flink.table.types.logical.utils.LogicalTypeCasts.supportsImplicitCast;
-
 /** Specific {@link ArgumentTypeStrategy} for {@link BuiltInFunctionDefinitions#ARRAY_CONTAINS}. */
 @Internal
 class ArrayElementArgumentTypeStrategy implements ArgumentTypeStrategy {
 
+    private final boolean preserveNullability;
+
+    public ArrayElementArgumentTypeStrategy(boolean preserveNullability) {
+        this.preserveNullability = preserveNullability;
+    }
+
     @Override
     public Optional<DataType> inferArgumentType(
             CallContext callContext, int argumentPos, boolean throwOnFailure) {
         final ArrayType haystackType =
                 (ArrayType) callContext.getArgumentDataTypes().get(0).getLogicalType();
         final LogicalType haystackElementType = haystackType.getElementType();
-        final LogicalType needleType =
-                callContext.getArgumentDataTypes().get(argumentPos).getLogicalType();
-        if (supportsImplicitCast(needleType, haystackElementType)) {
```

Comment:
I didn't get: check for implicit cast should be removed?

Reply:
Based on the Contract for `org.apache.flink.table.types.inference.ArgumentTypeStrategy#inferArgumentType` it should return an empty value in case input type could not be inferred. 
Right not this change violates this contract
https://github.com/apache/flink/blob/464ded1c2a0497255b70f711167c3b7ae52ea0f7/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/ArgumentTypeStrategy.java#L33-L48

Reply:
@snuyanzin but if do not change this, it will be always throw expcetion. Do you have any good suggestions？

Reply:
I'm pretty sure there should be a simpler way, however i can have a look closer to the end of the week

Reply:
hi @snuyanzin Since array_remove depends on this work, could you please help to confirm it as soon as possible?

Reply:
hi @snuyanzin ,do you have time to do?

Reply:
`haystackElementType` should be nullable or not depending on `needle` type nullability.
Then `supportsImplicitCast` will keep working.

Reply:
@snuyanzin i see. I misunderstood supportsImplicitCast before. have fixed this